# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: widget/nsDragServiceProxy.cpp
# Commit: 802512ee57ae
# Full Hash: 802512ee57aec9d975ab078c48b9f37956012bc6
# Author: Neil Deakin <neil@mozilla.com>
# Date: 2015-04-17 19:45:37
# Regressor Bug: 1071562
# File Overlap Count: 1
# Description:
#   Bug 1071562, e10s, support non-text types in clipboard (html, images, etc), r=smaug
# ==============================================================================

diff -r ffd1b16f058b -r 802512ee57ae widget/nsDragServiceProxy.cpp
--- a/widget/nsDragServiceProxy.cpp	Thu Apr 16 01:11:05 2015 +0300
+++ b/widget/nsDragServiceProxy.cpp	Thu Apr 16 15:38:12 2015 -0400
@@ -58,30 +58,19 @@
     if (surface) {
       mozilla::RefPtr<mozilla::gfx::DataSourceSurface> dataSurface =
         surface->GetDataSurface();
-      mozilla::gfx::DataSourceSurface::MappedSurface map;
-      dataSurface->Map(mozilla::gfx::DataSourceSurface::MapType::READ, &map);
       mozilla::gfx::IntSize size = dataSurface->GetSize();
-      mozilla::CheckedInt32 requiredBytes =
-        mozilla::CheckedInt32(map.mStride) * mozilla::CheckedInt32(size.height);
-      size_t maxBufLen = requiredBytes.isValid() ? requiredBytes.value() : 0;
-      mozilla::gfx::SurfaceFormat format = dataSurface->GetFormat();
-      // Surface data handling is totally nuts. This is the magic one needs to
-      // know to access the data.
-      size_t bufLen =
-        maxBufLen - map.mStride + (size.width * BytesPerPixel(format));
 
-      // nsDependentCString wants null-terminated string.
-      mozilla::UniquePtr<char[]> dragImageData(new char[maxBufLen + 1]);
-      memcpy(dragImageData.get(), reinterpret_cast<char*>(map.mData), bufLen);
-      memset(dragImageData.get() + bufLen, 0, maxBufLen - bufLen + 1);
-      nsDependentCString dragImage(dragImageData.get(), maxBufLen);
+      size_t length;
+      int32_t stride;
+      mozilla::UniquePtr<char[]> surfaceData =
+        nsContentUtils::GetSurfaceData(dataSurface, &length, &stride);
+      nsDependentCString dragImage(surfaceData.get(), length);
 
       mozilla::unused <<
         child->SendInvokeDragSession(dataTransfers, aActionType, dragImage,
-                                     size.width, size.height, map.mStride,
-                                     static_cast<uint8_t>(format),
+                                     size.width, size.height, stride,
+                                     static_cast<uint8_t>(dataSurface->GetFormat()),
                                      dragRect.x, dragRect.y);
-      dataSurface->Unmap();
       StartDragSession();
       return NS_OK;
     }
